### PR TITLE
crop and rotate : allow float/real cropping format ratios at input

### PR DESCRIPTION
**Pull request by aurelienpierre**
1. treat 5:4 or 4:5 the same way for consistency since, the portrait/landscape orientation doesn't depend on format ratio,
2. always reduce fractions using the greatest common divider, so inputting 6:4 is equivalent to 3:2,
3. allow real format ratios, so inputting 1.5 is equivalent to 0.6667 or 3:2 or 2:3, by finding the closest rational fraction,
4. search if the custom input ratio matches a default one and update the label accordingly.
5. unify the format ratio combobox update with a single path.

### CODE REVIEW

File: src/iop/clipping.c [hunk 1]
```diff
@@ -1798,22 +1924,6 @@ void gui_update(struct dt_iop_module_t *self)
...
-  {
-    const dt_iop_clipping_aspect_t *aspect = iter->data;
-    if((aspect->d == d) && (aspect->n == n))
-    {
-      act = i;
-      break;
-    }
-    i++;
-    iter = g_list_next(iter);
-  }
```
Comment (aurelienpierre) [inline src/iop/clipping.c:1816]: this is now gone in `aspect_presets_changed`

File: src/iop/clipping.c [hunk 2]
```diff
@@ -1828,20 +1938,7 @@ void gui_update(struct dt_iop_module_t *self)
...
-  if(act == -1)
-  {
-    char str[128];
-    snprintf(str, sizeof(str), "%d:%d", abs(p->ratio_d), p->ratio_n);
-    dt_bauhaus_combobox_set_text(g->aspect_presets, str);
-  }
-  if(dt_bauhaus_combobox_get(g->aspect_presets) == act)
-    aspect_presets_changed(g->aspect_presets, self);
-  else
-    dt_bauhaus_combobox_set(g->aspect_presets, act);
```
Comment (aurelienpierre) [inline src/iop/clipping.c:1844]: this is now gone in `aspect_presets_changed`, and the case where `dt_bauhaus_combobox_get(g->aspect_presets) == act` is now a no-op.

File: src/iop/clipping.c [hunk 3]
```diff
@@ -1615,9 +1616,101 @@ static void aspect_presets_changed(GtkWidget *combo, dt_iop_module_t *self)
...
+      {
+        // find the closest fraction from the input ratio
+        float rr = atof(text);
+        int dd = ceilf(rr);
+        int nn = floor(rr);
+
+        // some sanity check
+        if(dd == 0)
+        {
+          dt_control_log(_("invalid ratio format. it should be greater than one"));
```
Comment (parafin) [inline src/iop/clipping.c:1632]: greater than zero? or just "positive"

Comment (parafin) [reply src/iop/clipping.c:1632]: Also it doesn't look like you're checking for negative values either.

Comment (aurelienpierre) [reply src/iop/clipping.c:1632]: ooops, good catch.

Comment (aurelienpierre) [reply src/iop/clipping.c]: the only blocking case is if ==0, negative values are not an issue.

Comment (aurelienpierre) [reply src/iop/clipping.c]: edit : negative values make it joyfully crash, but that wasn't handled in the previous code either.

**Comment by TurboGit** [comment]
Works for me. Thanks!

**Pull request merged**